src/saas_shield/deterministic.rs

```diff
@@ -175,6 +161,83 @@ impl DeterministicFieldOps for SaasShieldDeterministicClient {
             .collect()
     }
 
+    async fn rotate_fields(
+        &self,
+        encrypted_fields: EncryptedFields,
+        metadata: &AlloyMetadata,
+        new_tenant_id: TenantId,
+    ) -> Result<RotateBatchResult, AlloyError> {
+        let paths = encrypted_fields
+            .values()
+            .map(|field| (field.secret_path.clone(), field.derivation_path.clone()))
+            .collect_vec();
+        let original_tenant_keys = derive_keys_many_paths(
+            &self.tenant_security_client,
+            metadata,
+            paths.clone(),
+            SecretType::Deterministic,
+        )
+        .await?;
+        let new_tenant_keys = if new_tenant_id != metadata.tenant_id {
+            derive_keys_many_paths(
+                &self.tenant_security_client,
+                metadata,
+                paths,
+                SecretType::Deterministic,
+            )
+            .await?
+        } else {
+            original_tenant_keys.clone() // TODO: I only need a reference to this, but the `else` creates an owned value.
+                                         // Is there a better way to do it?
+        };
+        let reencrypt_field = |encrypted_field: EncryptedField| {
+            let (original_key_id, ciphertext) =
+                Self::decompose_encrypted_field_header(encrypted_field.encrypted_field.clone())?;
+            let original_key = derive_key_for_path(
+                &original_tenant_keys,
+                &encrypted_field.secret_path,
+                &encrypted_field.derivation_path,
+                DeriveKeyChoice::Specific(original_key_id),
+            )?;
+            let new_current_key = derive_key_for_path(
```

Comment:
This reads a bit weird because you derive keys even if we don't actually have to do the rotation.

If we added a function on `KeyDeriveResponse` that takes `secret_path` and `derivation_path` and returns `Option<KeyId>` of the current key id for that path we could use that instead.

Reply:
I feel like the issue here is more in the naming. We call `derive_keys_many_paths` above to make a single call to the TSP and derive all the keys. Then this `derive_key_for_path` is badly named because it's not deriving, it's just getting things out of the HashMap based on the `DeriveKeyChoice` (in this case `Current`). 

We almost already have the function you're talking about on `KeyDeriveResponse`, which is what `derive_key_for_path` calls, it just also adds a consistent error message for the `None` case. But limiting the function to only return the KeyId instead of the whole `DerivedKey` when we also need the bytes doesn't seem worth it to me.

Reply:
Made some changes here in the end. Lmk what you think